### PR TITLE
pip 24, setuptools 69.5.1, hatchling 1.22.5, hatch_fancy_pypi_readme 24.1.0, platformdirs 4.2.0, packaging 24.0, trove_classifiers 2024.4.10, wheel 0.43.0

### PR DESCRIPTION
**Pull request by mkoeppe**
### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by mkoeppe** [comment]
There is a strange malfunction on ubuntu-mantic-standard
https://github.com/sagemath/sage/actions/runs/7852863047/job/21431835515?pr=37277#step:11:10319
```
sage -t --random-seed=293799319091173329856561880001578538606 src/sage/misc/package.py
**********************************************************************
File "src/sage/misc/package.py", line 186, in sage.misc.package.pip_installed_packages
Failed example:
    d['prompt-toolkit']
Exception raised:
    Traceback (most recent call last):
      File "/sage/src/sage/doctest/forker.py", line 712, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/sage/src/sage/doctest/forker.py", line 1147, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.misc.package.pip_installed_packages[3]>", line 1, in <module>
        d['prompt-toolkit']
        ~^^^^^^^^^^^^^^^^^^
    KeyError: 'prompt-toolkit'
**********************************************************************
```

`pip list` sees `prompt_toolkit` (underscore) instead `prompt-toolkit` (dash)

**Comment by mkoeppe** [comment]
The build failures in CI Linux are unrelated (#37786)

**Comment by kwankyu** [comment]
> There is a strange malfunction on ubuntu-mantic-standard https://github.com/sagemath/sage/actions/runs/7852863047/job/21431835515?pr=37277#step:11:10319
> 
> ```
> sage -t --random-seed=293799319091173329856561880001578538606 src/sage/misc/package.py
> **********************************************************************
> File "src/sage/misc/package.py", line 186, in sage.misc.package.pip_installed_packages
> Failed example:
>     d['prompt-toolkit']
> Exception raised:
>     Traceback (most recent call last):
>       File "/sage/src/sage/doctest/forker.py", line 712, in _run
>         self.compile_and_execute(example, compiler, test.globs)
>       File "/sage/src/sage/doctest/forker.py", line 1147, in compile_and_execute
>         exec(compiled, globs)
>       File "<doctest sage.misc.package.pip_installed_packages[3]>", line 1, in <module>
>         d['prompt-toolkit']
>         ~^^^^^^^^^^^^^^^^^^
>     KeyError: 'prompt-toolkit'
> **********************************************************************
> ```
> 
> `pip list` sees `prompt_toolkit` (underscore) instead `prompt-toolkit` (dash)
> 
> But this build is incremental from an ancient version:
> 
> ```
> ### Incremental build from SageMath version 10.1.beta6, Release Date: 2023-07-09
> ```
> 
> so can probably ignore.

I get the same failure with sage built from scratch on mac.

**Comment by kwankyu** [comment]
Both `pip install prompt_toolkit` and `pip install prompt-toolkit` work, and we have
```
upstream_url=https://pypi.io/packages/source/p/prompt_toolkit/prompt_toolkit-VERSION.tar.gz  
```
Why `prompt-tookit` (with dash) is expected?

**Comment by mkoeppe** [comment]
> Both `pip install prompt_toolkit` and `pip install prompt-toolkit` work

Yes, these are all equivalent.

I see the failure now on various platforms. e.g. https://github.com/mkoeppe/sage/actions/runs/8761823524/job/24053507902#step:11:12409 

I'll investigate.

File: src/sage/misc/package.py [hunk 1]
```diff
@@ -181,10 +181,10 @@ def pip_installed_packages(normalization=None):
...
         True
-        sage: d['beautifulsoup4']                               # optional - beautifulsoup4
-        '...'
-        sage: d['prompt-toolkit']
-        '...'
+        sage: sorted(d)                                         # needs beautifulsoup4
+        [...'beautifulsoup4'...]
+        sage: sorted(d)  # whether - or _ appears in the name depends on tool versions
+        [...'prompt...toolkit'...]
         sage: d = pip_installed_packages(normalization='spkg')
```
Comment (kwankyu) [inline src/sage/misc/package.py:188]: How about
```
sage: 'beautifulsoup4' in d
True
sage: 'prompt-toolkit' in d or 'prompt_toolkit' in d   # whether ...
True
```
?

Comment (kwankyu) [reply src/sage/misc/package.py:188]: Which "tool" do you mean? `setuptools`?

Comment (mkoeppe) [reply src/sage/misc/package.py:188]: or maybe `pip`...

Comment (kwankyu) [reply src/sage/misc/package.py:188]: If it's not clear, 
```
sage: 'prompt-toolkit' in d or 'prompt_toolkit' in d  
True
```
may suffice.

Comment (kwankyu) [reply src/sage/misc/package.py:188]: Thanks.

**Review by kwankyu — APPROVED** [review]
LGTM.

**Comment by kwankyu** [comment]
Good. Thanks.

**Comment by mkoeppe** [comment]
Thanks for the review!

**Pull request merged**